File: go/irg.go
// Package irg implements a parser for the TC004 IRG thermal image format.
package main

import (
	"encoding/binary"
	"fmt"
	"io"
	"math"
	"os"
)

// IrgHeader contains parsed metadata from an IRG file header.
type IrgHeader struct {
	UnknownHeader          int32
	FirstImageSize         uint32
	FirstImageWidth        uint16
	FirstImageHeight       uint16
	Pad1                   int8
	SecondImageSize        uint32
	SecondImageWidth       uint16
	SecondImageHeight      uint16
	Pad2                   uint8
	ThirdImageSize         uint32
	ThirdImageWidth        uint16
	ThirdImageHeight       uint16
	Emissivity             uint32
	ReflectiveTemperature  uint32
	AmbientTemperature     uint32
	Distance               uint32
	Unknown                uint32
	Transmissivity32       uint32
	Padding                uint32
	Transmissivity         uint16
}

// IrgData holds all extracted data from an IRG file.
type IrgData struct {
	Header           IrgHeader
	Grayscale        []uint8
	ThermalRaw       []uint16
	ThermalWidth     int
	ThermalHeight    int
}

// ThermalFahrenheit converts raw thermal data to Fahrenheit.
func (d *IrgData) ThermalFahrenheit() []float64 {
	result := make([]float64, len(d.ThermalRaw))
	for i, v := range d.ThermalRaw {
		kelvin := float64(v) / 10.0
		result[i] = (kelvin-273.15)*9.0/5.0 + 32.0
	}
	return result
}

// ThermalMinMax returns the min and max Fahrenheit values.
func (d *IrgData) ThermalMinMax() (float64, float64) {
	f := d.ThermalFahrenheit()
	if len(f) == 0 {
		return 0, 0
	}
	min, max := math.Inf(1), math.Inf(-1)
	for _, v := range f {
		if v < min {
			min = v
		}
		if v > max {
			max = v
		}
	}
	return min, max
}

// ExtractIRG parses an IRG file and returns extracted image data.
func ExtractIRG(path string) (*IrgData, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("open file: %w", err)
	}
	defer f.Close()

	// Read entire file
	data, err := io.ReadAll(f)
	if err != nil {
		return nil, fmt.Errorf("read file: %w", err)
	}
	if len(data) < 0x100 {
		return nil, fmt.Errorf("file too small to be a valid IRG file (%d bytes)", len(data))
	}

	// Parse header using a struct that matches the binary layout.
	// The Python format string is: '<iIHHbIHHBIHHIIIIIIIH14xB'
	// We parse field-by-field for clarity.
	var hdr IrgHeader
	hdr.UnknownHeader = int32(binary.LittleEndian.Uint32(data[0x00:]))
	hdr.FirstImageSize = binary.LittleEndian.Uint32(data[0x04:])
	hdr.FirstImageWidth = binary.LittleEndian.Uint16(data[0x08:])
	hdr.FirstImageHeight = binary.LittleEndian.Uint16(data[0x0A:])
	hdr.Pad1 = int8(data[0x0C])
	hdr.SecondImageSize = binary.LittleEndian.Uint32(data[0x0D:])
	hdr.SecondImageWidth = binary.LittleEndian.Uint16(data[0x11:])
	hdr.SecondImageHeight = binary.LittleEndian.Uint16(data[0x13:])
	hdr.Pad2 = data[0x15]
	hdr.ThirdImageSize = binary.LittleEndian.Uint32(data[0x16:])
	hdr.ThirdImageWidth = binary.LittleEndian.Uint16(data[0x1A:])
	hdr.ThirdImageHeight = binary.LittleEndian.Uint16(data[0x1C:])
	hdr.Emissivity = binary.LittleEndian.Uint32(data[0x1E:])
	hdr.ReflectiveTemperature = binary.LittleEndian.Uint32(data[0x22:])
	hdr.AmbientTemperature = binary.LittleEndian.Uint32(data[0x26:])
	hdr.Distance = binary.LittleEndian.Uint32(data[0x2A:])
	hdr.Unknown = binary.LittleEndian.Uint32(data[0x2E:])
	hdr.Transmissivity = binary.LittleEndian.Uint16(data[0x32:])

	// Determine data start offset
	dataStart := 0x100
	if data[0x7E] == 0xAC && data[0x7F] == 0xCA {
		dataStart = 0x80
	}

	offset := dataStart

	// Extract grayscale image (uint8 array)
	gsSize := int(hdr.FirstImageSize)
	if offset+gsSize > len(data) {
		return nil, fmt.Errorf("file too small for grayscale data")
	}
	grayscale := make([]uint8, gsSize)
	copy(grayscale, data[offset:offset+gsSize])
	offset += gsSize

	// Extract thermal data (uint16 array)
	thermalCount := int(hdr.SecondImageSize)
	thermalByteSize := thermalCount * 2
	if offset+thermalByteSize > len(data) {
		return nil, fmt.Errorf("file too small for thermal data")
	}
	thermalRaw := make([]uint16, thermalCount)
	for i := 0; i < thermalCount; i++ {
		thermalRaw[i] = binary.LittleEndian.Uint16(data[offset+i*2:])
	}

	return &IrgData{
		Header:        hdr,
		Grayscale:     grayscale,
		ThermalRaw:    thermalRaw,
		ThermalWidth:  int(hdr.SecondImageWidth),
		ThermalHeight: int(hdr.SecondImageHeight),
	}, nil
}

File: rust/src/irg.rs
use std::fs;
use std::io::{self, Cursor, Read};

/// Parsed header fields from an IRG file.
#[derive(Debug)]
pub struct IrgHeader {
    pub unknown_header: i32,
    pub first_image_size: u32,
    pub first_image_width: u16,
    pub first_image_height: u16,
    pub second_image_size: u32,
    pub second_image_width: u16,
    pub second_image_height: u16,
    pub third_image_size: u32,
    pub third_image_width: u16,
    pub third_image_height: u16,
    pub emissivity: u32,
    pub reflective_temperature: u32,
    pub ambient_temperature: u32,
    pub distance: u32,
    pub transmissivity: u16,
}

/// Extracted image data from an IRG file.
pub struct IrgData {
    pub header: IrgHeader,
    pub grayscale: Vec<u8>,
    pub thermal_raw: Vec<u16>,
    pub thermal_width: usize,
    pub thermal_height: usize,
}

impl IrgData {
    /// Return thermal data converted to Fahrenheit.
    pub fn thermal_fahrenheit(&self) -> Vec<f64> {
        self.thermal_raw
            .iter()
            .map(|&v| {
                let kelvin = f64::from(v) / 10.0;
                (kelvin - 273.15) * 9.0 / 5.0 + 32.0
            })
            .collect()
    }
}

/// Read a little-endian u16 from a cursor.
fn read_u16_le(cur: &mut Cursor<&[u8]>) -> io::Result<u16> {
    let mut buf = [0u8; 2];
    cur.read_exact(&mut buf)?;
    Ok(u16::from_le_bytes(buf))
}

/// Read a little-endian u32 from a cursor.
fn read_u32_le(cur: &mut Cursor<&[u8]>) -> io::Result<u32> {
    let mut buf = [0u8; 4];
    cur.read_exact(&mut buf)?;
    Ok(u32::from_le_bytes(buf))
}

/// Read a little-endian i32 from a cursor.
fn read_i32_le(cur: &mut Cursor<&[u8]>) -> io::Result<i32> {
    let mut buf = [0u8; 4];
    cur.read_exact(&mut buf)?;
    Ok(i32::from_le_bytes(buf))
}

/// Parse an IRG file and extract all image data.
pub fn extract_irg(path: &str) -> io::Result<IrgData> {
    let bytes = fs::read(path)?;
    if bytes.len() < 0x100 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "File too small to be a valid IRG file",
        ));
    }

    let mut cur = Cursor::new(bytes.as_slice());

    // Parse header (matches the Python struct format '<iIHHbIHHBIHHIIIIIIIH14xB')
    let unknown_header = read_i32_le(&mut cur)?;
    let first_image_size = read_u32_le(&mut cur)?;
    let first_image_width = read_u16_le(&mut cur)?;
    let first_image_height = read_u16_le(&mut cur)?;
    // pad1 (1 byte)
    let mut _pad = [0u8; 1];
    cur.read_exact(&mut _pad)?;

    let second_image_size = read_u32_le(&mut cur)?;
    let second_image_width = read_u16_le(&mut cur)?;
    let second_image_height = read_u16_le(&mut cur)?;
    // pad2 (1 byte)
    cur.read_exact(&mut _pad)?;

    let third_image_size = read_u32_le(&mut cur)?;
    let third_image_width = read_u16_le(&mut cur)?;
    let third_image_height = read_u16_le(&mut cur)?;

    let emissivity = read_u32_le(&mut cur)?;
    let reflective_temperature = read_u32_le(&mut cur)?;
    let ambient_temperature = read_u32_le(&mut cur)?;
    let distance = read_u32_le(&mut cur)?;
    let _unknown = read_u32_le(&mut cur)?;
    let _transmissivity_u32 = read_u32_le(&mut cur)?; // padding/unknown
    let _padding = read_u32_le(&mut cur)?;
    let transmissivity = read_u16_le(&mut cur)?;

    let header = IrgHeader {
        unknown_header,
        first_image_size,
        first_image_width,
        first_image_height,
        second_image_size,
        second_image_width,
        second_image_height,
        third_image_size,
        third_image_width,
        third_image_height,
        emissivity,
        reflective_temperature,
        ambient_temperature,
        distance,
        transmissivity,
    };

    // Determine data start offset
    let data_start = if bytes[0x7E] == 0xAC && bytes[0x7F] == 0xCA {
        0x80usize
    } else {
        0x100usize
    };

    let mut offset = data_start;

    // Extract grayscale image (uint8 array)
    let gs_size = first_image_size as usize;
    if offset + gs_size > bytes.len() {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "File too small for grayscale data",
        ));
    }
    let grayscale = bytes[offset..offset + gs_size].to_vec();
    offset += gs_size;

    // Extract thermal data (uint16 array, 2 bytes per element)
    let thermal_count = second_image_size as usize;
    let thermal_byte_size = thermal_count * 2;
    if offset + thermal_byte_size > bytes.len() {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "File too small for thermal data",
        ));
    }
    let thermal_raw: Vec<u16> = bytes[offset..offset + thermal_byte_size]
        .chunks_exact(2)
        .map(|c| u16::from_le_bytes([c[0], c[1]]))
        .collect();

    Ok(IrgData {
        header,
        grayscale,
        thermal_raw,
        thermal_width: second_image_width as usize,
        thermal_height: second_image_height as usize,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_read_u16_le() {
        let data = [0x34, 0x12];
        let mut cur = Cursor::new(data.as_slice());
        assert_eq!(read_u16_le(&mut cur).unwrap(), 0x1234);
    }

    #[test]
    fn test_read_u32_le() {
        let data = [0x78, 0x56, 0x34, 0x12];
        let mut cur = Cursor::new(data.as_slice());
        assert_eq!(read_u32_le(&mut cur).unwrap(), 0x12345678);
    }
}
